default_gateway.go

```diff
@@ -0,0 +1,158 @@
+package libnetwork
+
+import (
+	"fmt"
+
+	"github.com/docker/libnetwork/netlabel"
+	"github.com/docker/libnetwork/options"
+	"github.com/docker/libnetwork/types"
+)
+
+const (
+	libnGWNetwork = "GW_bridge"
+)
+
+/*
+   libnetwork creates a bridge network "GW_bridge" for provding
+   default gateway for the containers if none of the container's
+   endpoints have GW set by the driver. ICC is set to false for
+   the GW_bridge network.
+
+   If a driver can't provide external connectivity it can choose
+   to not set the GW IP for the endpoint.
+
+   endpoint on the GW_bridge network is managed dynamically by
+   libnetwork. ie:
+   - its created when an endpoint without GW joins the container
+   - its deleted when an endpoint with GW joins the container
+*/
+
+func (sb *sandbox) setupDefaultGW(srcEp *endpoint) error {
+	var createOptions []EndpointOption
+	c := srcEp.network.ctrlr
+
+	// check if the conitainer already has a GW endpoint
+	if ep := sb.getEndpointInGWNetwork(); ep != nil {
+		return nil
+	}
+
+	n, err := c.NetworkByName(libnGWNetwork)
+	if err != nil {
+		if _, ok := err.(ErrNoSuchNetwork); !ok {
+			return err
+		}
+		n, err = createGWNetwork(c)
+		if err != nil {
+			return err
+		}
+	}
+
+	if opt, ok := srcEp.generic[netlabel.PortMap]; ok {
+		if pb, ok := opt.([]types.PortBinding); ok {
+			createOptions = append(createOptions, CreateOptionPortMapping(pb))
+		}
+	}
+
+	if opt, ok := srcEp.generic[netlabel.ExposedPorts]; ok {
+		if exp, ok := opt.([]types.TransportPort); ok {
+			createOptions = append(createOptions, CreateOptionExposedPorts(exp))
+		}
+	}
+
+	newEp, err := n.CreateEndpoint(srcEp.name, createOptions...)
```

Comment:
This endpoint is really created to provide default gw for this specific sandbox. When the sandbox is destroyed or when it no longer needs this auxiliary endpoint, this endpoint will be deleted.
Given this, I feel this endpoint name should somehow embed this sandbox identifier.
I think it would help in case of debugging.